### PR TITLE
`Integrated code lifecycle`: Limit build jobs that can be queued per participation

### PR DESCRIPTION
**Pull request by BBesrour**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.

### Motivation and Context

Currently, users can spam submissions and occupy the build job queue. We need to implement a mechanism that checks, whether build jobs in the queue (not the running jobs) for that specific repository already exist. The job needs to keep its original position in the queue

### Description

Changed that the queue only holds a reference of the build job. The actual build job is stored in a Hazelcast map. **We use participationID as the key**. If a user submits multiple times, the queue will still hold multiple entries for that participation, but only the first item in the queue will actually result in a build job (since we remove the item from the map before we process it).
I considered these 3 approaches:
- Iterating through the queue to check for existing build jobs for the same participation: This is not ideal since the hazelcast queue aren't suitable for replacing items in the queue
- Replacing the queue with a map: Since we prioritize specific build jobs according to some criteria, we would have to get the map everytime we need to process a job and sort accordingly. this isn't ideal as this part needs to be synchronized between all the build agents and it would hurt performance
- Add a map that holds the actual build job and keep the queue to take care of the prioritization

### Steps for Testing

Test locally or on a test server with admin rights
If testing locally, add this to your `application-local.yml` as it will make testing easier
```
artemis:
    continuous-integration:
        specify-concurrent-builds: true
        concurrent-build-size: 1
```
Prerequisites:
- 1 Admin
- 1 Student
- 2 Programming Exercise

0. Start Artemis with integrated code lifecycle
1. As the admin, open the Build Queue (Server Administration > Build Queue). Always keep this open in a tab where you can see it.
2. As the student, open the 2 programming exercises in two different tabs,
3. Spam the submit button on both exercises. The running build jobs will be filled and should see a **maximum of 2 build jobs queued at any time**
4. Make sure that the build job result displayed at the end is linked with the latest commit.

#### Performance Review
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **Refactor**
	- Renamed `BuildJobQueueItem` to `BuildJobItem` across various modules for consistency.
	- Introduced `BuildJobItemReference` and `BuildJobItemReferenceDTO` for efficient build job item management.
	- Updated data structures and method signatures to utilize the new `BuildJobItem` naming.

- **New Features**
	- Added efficient management of build job items using `CircularFifoQueue`.

- **Bug Fixes**
	- Improved build job item handling logic to enhance performance and reliability.

### CODE REVIEW

**Review by florian-glombik — APPROVED** [review]
Tested locally in testing session for normal exercises (#build-jobs per participation set to 1), the following scenarios worked fine:

- pushing comments, pushing a build error -> build error as result
- pushing comments, pushing build error, pushing fix for build error -> no build error as result

The queue displayed only 1 running job as expected.

**Review by muradium — DISMISSED** [review]
Tested during testing session, queue had only one build job per submission as expected

**Review by SimonEntholzer — APPROVED** [review]
Code looks good, the renaming makes sense in my opinion. 

Also tested locally, the queue indeed doesn't fill up to more than 2 jobs (for 2 exercises submitted by one user), compared to multiple jobs piling up at the current develop, when spamming submit.

**Review by laurenzfb — APPROVED** [review]
code lgtm

**Review by florian-glombik — APPROVED** [review]
Re-tested the following scenarios for normal exercises (#build-jobs per participation set to 1):

- pushing comments, pushing a build error -> build error as result
- pushing comments, pushing build error, pushing fix for build error -> no build error as result

The queue displayed only 1 running job as expected.

The reference to the commit hash was changed for the queued elements -> spamming the submit button does not queue multiple build jobs

**Review by SimonEntholzer — DISMISSED** [review]
re-approve

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by muradium — APPROVED** [review]
Reapproval

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge

**Review by SimonEntholzer — APPROVED** [review]
Re-approve after merge

**Review by florian-glombik — APPROVED** [review]
Re-Approve after merge

**Comment by BBesrour** [comment]
Requirements changed to allow a configurable number of jobs per participation. I will have to adjust the implementation.
This PR will be delayed due to higher priority tasks.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/dto/BuildJobItemReference.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/dto/BuildJobItemReference.java:18]:
```suggestion
public record BuildJobItemReferenceDTO(String id, long participationId, long courseId, int priority, ZonedDateTime submissionDate) implements Serializable {
```
The server test currently fail due to this

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/SharedQueueProcessingService.java
Comment (BBesrour) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/SharedQueueProcessingService.java:74]: Where is `String` used as key for buildJobItemMap

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/SharedQueueProcessingService.java
Comment (BBesrour) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/SharedQueueProcessingService.java:225]: How?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCITriggerService.java
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCITriggerService.java:83]: My first impression when reading `max-build-job-items` is that this limits concurrency for the whole system, not just per participation. Can you find a better name for this configuration parameter? Maybe `parallel-jobs-per-participation`.

**Review by marlon-luca-bu — DISMISSED** [review]
Code

**Review by yassinsws — DISMISSED** [review]
Code Lgtm

**Review by iyannsch — APPROVED** [review]
Tested in Testing Session on 12.06. - looks good :)

**Review by az108 — COMMENTED** [review]
Tested in Testing Session on 2 Test Servers with explanation and running concurrent builds and verifying the behaviour

**Review by az108 — DISMISSED** [review]

**Review by kaancayli — DISMISSED** [review]
Tested during testing session. Code looks good. Works as expected

**Review by coolchock — APPROVED** [review]
tested during testing session, works as expected

**Review by JohannesStoehr — COMMENTED** [review]
As far as I can see the buildJobItemMap is created automatically by Hazelcast. I can only find deletions if all build jobs are cancelled, so the map should automatically clean up other old entries. Is that the case?
Did you verify that the map will not delete entries under high load, e.g. during an exam with 2000 students?

**Review by iyannsch — APPROVED** [review]
Code LGTM

**Pull request closed without merging**
